### PR TITLE
Fixes #3449 — Add The Ability To View User Referrals

### LINKED ISSUE #3449: Add The Ability To View User Referrals To Member Profile
Per a discussion on Discord, currently the member profile shows only the number of referrals without a way to view the users individually, but should allow the user to view a list of all user referrals.

### PR DESCRIPTION
**Pull request by WildcardSearch**
Adds the ability to view referrals where applicable throughout the project.

**Changes:**
- edited two (2) templates (`member_profile_referrals` and `headerinclude`)
- added six (6) templates (`member_no_referrals`, `member_profile_referrals`, `member_referral_row`, `member_referrals`, `member_referrals_link`, and `member_referrals_popup`)
- added one (1) setting
- added three (3) plugin hooks (`member_referrals_start`, `member_referrals_end`, and `xmlhttp_referrals_end`)
- added one setting peeker to toggle `referralsperpage` per `usereferrals`

Modified referral display to show `0` when there are no referrals for the user and show the referral count linked to a forum page to display the referrals as a fallback for JS. Uses JS to prevent linking and instead displays a modal with the referral list.

Modifies display of referrals on:
- member profile
- usercp overview
- member list

**note: to test the fallback forum page, right-click the referral link and open the link directly or in a new tab/window**

Fixes #3449

### CODE REVIEW

**Comment by Eldenroot** [comment]
Works fine, thank you

**Comment by WildcardSearch** [comment]
Thanks, @Eldenroot 

If no one has any issues with it in the next few days we should be able to merge it.

File: inc/functions.php [hunk 1]
```diff
@@ -8929,4 +8929,50 @@ function my_hash_equals($known_string, $user_string)
 
 		return $result === 0;
 	}
+}
+
+/**
+ * Retrieves all referrals for a specified user
+ *
+ * @param init uid
```
Comment (euantorano) [inline inc/functions.php]: Should be `int`, not `init`.

File: inc/functions.php [hunk 2]
```diff
@@ -8929,4 +8929,50 @@ function my_hash_equals($known_string, $user_string)
...
+ * @param int start position
+ * @param int total entries
+ * @param bool false (default) only return display info, true for all info
+ * @return bool Result of the comparison.
```
Comment (euantorano) [inline inc/functions.php]: Actually returns an `array`.

File: inc/functions.php [hunk 3]
```diff
@@ -8929,4 +8929,50 @@ function my_hash_equals($known_string, $user_string)
...
+		$query_options['limit'] = $limit;
+	}
+
+	$fields = 'uid, username, usergroup, displaygroup, regdate';
+	if($full === true)
+	{
+		$fields = '*';
+	}
+
+	$query = $db->simple_select('users', $fields, "referrer='{$uid}'", $query_options);
```
Comment (euantorano) [inline inc/functions.php]: Might be best to cast `$uid` just to be sure.

**Comment by WildcardSearch** [comment]
Done, @euantorano

**Pull request merged**